### PR TITLE
Deprecate cell() in favor of Array{Any}()

### PR DESCRIPTION
**Pull request by nalimilan**
This MATLAB-ism feels out of place in Julia, and isn't used very often
in the code base. This change also makes the names of C functions more
consistent with other ones.

(Note that most of the changes happen in C code, and aren't strictly needed to clean the Julia API.)

### CODE REVIEW

**Comment by kmsquire** [comment]
:+1:

File: base/markdown/Julia/interp.jl [hunk 1]
```diff
@@ -39,7 +39,7 @@ end
 
 toexpr(x) = x
 
-toexpr(xs::Vector{Any}) = Expr(:cell1d, map(toexpr, xs)...)
+toexpr(xs::Vector{Any}) = Expr(:vectany, map(toexpr, xs)...)
```
Comment (JeffBezanson) [inline base/markdown/Julia/interp.jl]: Since the syntax that parses to `cell1d` is deprecated, maybe it makes sense just to leave this until we have an official new meaning for it?

File: src/alloc.c [hunk 2]
```diff
@@ -327,8 +327,8 @@ void jl_lambda_info_set_ast(jl_lambda_info_t *li, jl_value_t *ast)
     jl_gc_wb(li, li->ssavaluetypes);
     int i;
     for(i=0; i < nslots; i++) {
-        jl_value_t *vi = jl_cellref(vis, i);
-        jl_sym_t *name = (jl_sym_t*)jl_cellref(vi, 0);
+        jl_value_t *vi = jl_array_any_ref(vis, i);
```
Comment (JeffBezanson) [inline src/alloc.c]: This function works for any pointer array, not just `Array{Any}`.

File: base/markdown/Julia/interp.jl (hunk 1 above)
Comment (nalimilan) [inline base/markdown/Julia/interp.jl]: Well, it's really easy to rename, so I figured it would be simpler to do this is small steps. (I actually tried to remove the code parsing the deprecated syntax, but it's hard when you don't understand the code...)

File: src/alloc.c (hunk 2 above)
Comment (nalimilan) [inline src/alloc.c]: OK, so it's the occasion to make this clearer. What's the best name for it? `jl_array_ptr_ref`?

**Comment by JeffBezanson** [comment]
Definitely :+1: to deprecating `cell`.

File: base/markdown/Julia/interp.jl (hunk 1 above)
Comment (JeffBezanson) [inline base/markdown/Julia/interp.jl]: Ok, I guess since this was deprecated in 0.4 we should actually remove it completely, and use `Expr(:ref, :Any, ...)` instead.

Comment (nalimilan) [inline base/markdown/Julia/interp.jl]: Yeah, but let's leave that for another PR, since I don't think I would be able to do that.

**Comment by nalimilan** [comment]
@JeffBezanson Is the new version OK?

File: src/jltypes.c [hunk 3]
```diff
@@ -3509,7 +3509,7 @@ void jl_init_types(void)
     jl_array_type->pointerfree = 0;
     jl_array_type->ninitialized = 0;
 
-    jl_array_any_type =
+    jl_array_ptr_type =
```
Comment (vtjnash) [inline src/jltypes.c]: this should still be `jl_array_any_type`

Comment (nalimilan) [inline src/jltypes.c]: Sorry, I'm not familiar with this code. I just fixed it.

I also changed `vector_ptr` to `array_ptr_1d` to be more consistent with existing functions. It's more verbose, but we should rename them all at once if we want shorter identifiers.

**Comment by JeffBezanson** [comment]
I changed some names to be more to my liking. The issue is that `jl_alloc_cell_1d` does make exactly a `Vector{Any}`, so it should have `any` in the name. For the other functions, using `ptr` is not ideal since we also have a `Ptr` type, but it's ok for now.

**Comment by nalimilan** [comment]
OK, fine with me, as I said I don't know that C code at all. I'll merge tomorrow then if nobody objects.

**Comment by nalimilan** [comment]
@JeffBezanson Argh, I think I overwrote your changes while rebasing against master. If that's the case, could you force-push them again?

**Comment by JeffBezanson** [comment]
I don't have access to the local branch where I made the changes. Could you just remove all the changes to `src` so we can merge the operative part of this?

**Comment by tkelman** [comment]
https://github.com/JuliaLang/julia/commit/8393fc38fa9cb720e68eeff7ff65e6e1b6f84e9d or https://github.com/JuliaLang/julia/commit/be7b48ae7abc9b49b4afa994641512c3cda76471 ?

**Comment by JeffBezanson** [comment]
https://github.com/JuliaLang/julia/commit/8393fc38fa9cb720e68eeff7ff65e6e1b6f84e9d

**Comment by nalimilan** [comment]
Thanks @tkelman! I've rebased that commit, should be good to go.
